From f2d0c5a87da7a146f49a5c83ec1733990024e267 Mon Sep 17 00:00:00 2001
From: Ander Conselvan de Oliveira <conselvan2@gmail.com>
Date: Fri, 27 Jul 2012 13:30:51 +0300
Subject: [PATCH 8/9] libdrm: Fix bogus increment of a property set object
 count

If drmModePropertySetAdd() is called with the following property pairs
it will mistakenly increase the object count:

  (4, 1), (5,1), (4, 2).

When adding the third pair, the new_obj variable would be set when prev
points to (4, 1) since it disregarded the value of the previous item.

This patch changes the logic so that the obj count is incresed only if
both the previous and next items have differing object ids.
---
 xf86drmMode.c | 16 ++++++----------
 1 file changed, 6 insertions(+), 10 deletions(-)

diff --git a/xf86drmMode.c b/xf86drmMode.c
index 0485113..d63f099 100644
--- a/xf86drmMode.c
+++ b/xf86drmMode.c
@@ -1192,10 +1192,8 @@ int drmModePropertySetAdd(drmModePropertySetPtr set,
 
 	/* keep it sorted by object_id and property_id */
 	while (prev->next) {
-		if (prev->next->object_id > object_id) {
-			new_obj = true;
+		if (prev->next->object_id > object_id)
 			break;
-		}
 
 		if (prev->next->object_id == object_id &&
 		    prev->next->property_id >= property_id)
@@ -1204,8 +1202,8 @@ int drmModePropertySetAdd(drmModePropertySetPtr set,
 		prev = prev->next;
 	}
 
-	if (!prev->next &&
-	    (prev == &set->list || prev->object_id != object_id))
+	if ((prev == &set->list || prev->object_id != object_id) &&
+	    (!prev->next || prev->next->object_id != object_id))
 		new_obj = true;
 
 	/* replace or add? */
@@ -1254,10 +1252,8 @@ int drmModePropertySetAddBlob(drmModePropertySetPtr set,
 
 	/* keep it sorted by object_id and property_id */
 	while (prev->next) {
-		if (prev->next->object_id > object_id) {
-			new_obj = true;
+		if (prev->next->object_id > object_id)
 			break;
-		}
 
 		if (prev->next->object_id == object_id &&
 		    prev->next->property_id >= property_id)
@@ -1266,8 +1262,8 @@ int drmModePropertySetAddBlob(drmModePropertySetPtr set,
 		prev = prev->next;
 	}
 
-	if (!prev->next &&
-	    (prev == &set->list || prev->object_id != object_id))
+	if ((prev == &set->list || prev->object_id != object_id) &&
+	    (!prev->next || prev->next->object_id != object_id))
 		new_obj = true;
 
 	/* replace or add? */
-- 
2.3.1

